component-parser/main.go

```diff
@@ -0,0 +1,318 @@
+		if err != nil {
+			return err
+		}
+		if d.IsDir() || d.Name() != "metadata.yaml" {
+			return nil
+		}
+
+		compName := filepath.Base(filepath.Dir(path))
+		if slices.Contains(supportedComponents, compName) {
+			files = append(files, path)
+			return nil
+		}
+
+		return nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return files, nil
+}
+
+// getSupportLevel gets the support level for the given category
+func getSupportLevel(stability *Stability, category string) string {
+	if contains(stability.Unmaintained, category) {
+		return "Unmaintained"
+	}
+	if contains(stability.Deprecated, category) {
+		return "Deprecated"
+	}
+	if contains(stability.Stable, category) {
+		return "Stable"
+	}
+	if contains(stability.Beta, category) {
+		return "Beta"
+	}
+	if contains(stability.Development, category) {
+		return "Development"
+	}
+	if contains(stability.Alpha, category) {
+		return "Alpha"
+	}
+	return "n/a"
+}
+
+// Helper function to check if a slice contains a specific category
+func contains(slice []string, item string) bool {
+	for _, v := range slice {
+		if v == item {
+			return true
+		}
+	}
+	return false
+}
+
+func processMetadataFiles(files []string) map[string][]*Metadata {
+	componentsByClass := make(map[string][]*Metadata)
+	for _, file := range files {
+		sg, err := parseYAMLFile(file)
+		if err != nil {
+			log.Printf("Failed to parse %s: %v\n", file, err)
+			continue
+		}
+		if sg.Type == "" {
+			log.Printf("Skipping invalid component in file %s: missing type\n", file)
+			continue
+		}
+		class := sg.Status.Class
+		componentsByClass[class] = append(componentsByClass[class], sg)
+	}
+	return componentsByClass
+}
+
+func printComponentTable(componentType, header string, components []*Metadata, getSupportFunc func(*Metadata) string) {
+	w := tabwriter.NewWriter(os.Stdout, 0, 0, 2, ' ', tabwriter.Debug)
+	fmt.Println(strings.Repeat("=", len(header)))
+	fmt.Println(componentType)
+	fmt.Println(strings.Repeat("=", len(header)))
+	fmt.Fprintln(w, header)
+
+	for _, component := range components {
+		support := getSupportFunc(component)
+		fmt.Fprintf(w, "%s\t%s\n", component.Type, support)
+	}
+	w.Flush()
+	fmt.Println()
+}
+
+func main() {
+	manifestPath := flag.String("m", "manifest.yaml", "Specify the path to the manifest.yaml file")
+	flag.Parse()
+
+	repos := []string{
+		"opentelemetry-collector",
+		"opentelemetry-collector-contrib",
+	}
+
+	pwd, err := os.Getwd()
+	if err != nil {
+		log.Fatalln(err)
+	}
+
+	metadataFiles := []string{}
+
+	for _, repo := range repos {
+		manifest, err := getDistributionConfig(*manifestPath)
+		if err != nil {
+			log.Fatalf("Error getting distribution config: %v", err)
+		}
+
+		if _, err := os.Stat(filepath.Join(pwd, repo)); os.IsNotExist(err) {
+			log.Printf("Repository %s does not exist, cloning...\n", repo)
+			err := exec.Command("git", "clone", fmt.Sprintf("https://github.com/open-telemetry/%s.git", repo)).Run()
```

Comment:
Do we actually need to clone those repos? Could we not use the github api instead?

Pointing to some `go.mod` file should be should be sufficient to extract the information from the `metadata.yaml` files?

Reply:
For me, this method is simpler and easier to follow. I don't think we need those kind of optimizations for this use case.

Reply:
The main thing I see here is that the current approach is more error prune and someone needs to maintain it.
Also using exec instead of simply querying some API endpoints feels to me like bad practice.

---
As a starting point, all details are available in this manifest file:
https://raw.githubusercontent.com/open-telemetry/opentelemetry-collector-releases/refs/heads/main/distributions/otelcol-contrib/manifest.yaml

When iterating on the manifest details, we manly have to construct two types of domains...
Example:
- https://raw.githubusercontent.com/open-telemetry/opentelemetry-collector/refs/tags/exporter/debugexporter/v0.112.0/exporter/debugexporter/metadata.yaml
**or**
- https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.112.0/exporter/alertmanagerexporter/metadata.yaml

Finally merge all the metadata and render it.

Reply:
> The main thing I see here is that the current approach is more error prune and someone needs to maintain it. 

I don't still see why this is "more error-prone". A call to the GitHub API will need to be maintained too. 

> Also using exec instead of simply querying some API endpoints feels to me like bad practice.

I would agree if this would be an application run in production systems where we need high availability or something like that and not just an script to get what components are supported or not as part of our release process. 

> Additonally we dont run into any issues, when the exec runs on mac or linux machines with different git versions or limitations on the executed processes.

It's just "git clone". You should have it configured on your machine to be able to work.

```
As a starting point, all details are available in this manifest file:
https://raw.githubusercontent.com/open-telemetry/opentelemetry-collector-releases/refs/heads/main/distributions/otelcol-contrib/manifest.yaml

When iterating on the manifest details, we manly have to construct two types of domains...
Example:

https://raw.githubusercontent.com/open-telemetry/opentelemetry-collector/refs/tags/exporter/debugexporter/v0.112.0/exporter/debugexporter/metadata.yaml
or
https://raw.githubusercontent.com/open-telemetry/opentelemetry-collector-contrib/refs/tags/v0.112.0/exporter/alertmanagerexporter/metadata.yaml
Finally merge all the metadata and render it.

```

I don't get what you want to explain with this. Please, you could you detail it more?